spec/unit/actions/app_delete_spec.rb

```diff
@@ -311,30 +307,31 @@ module VCAP::CloudController
 
               it 'returns some errors describing that the service bindings are being deleted asynchronously' do
                 expect { app_delete.delete(app_dataset) }.to raise_error(AppDelete::SubResourceError) do |err|
-                  expect(err.underlying_errors).to match_array(
-                    [
-                      AppDelete::AsyncBindingDeletionsTriggered.new(
+                  expect(err.underlying_errors.map(&:message)).to contain_exactly(
+                    "An operation for the service binding between app #{binding2.app.name} and service instance #{binding2.service_instance.name} is in progress.",
                         "An operation for the service binding between app #{binding1.app.name} and service instance #{binding1.service_instance.name} is in progress."
-                      ),
-                      AppDelete::AsyncBindingDeletionsTriggered.new(
-                        "An operation for the service binding between app #{binding2.app.name} and service instance #{binding2.service_instance.name} is in progress."
-                      )
-                    ]
                   )
                 end
               end
             end
           end
 
           context 'when service binding delete returns errors' do
+            let!(:binding1) { ServiceBinding.make(app: app, service_instance: ManagedServiceInstance.make(space: app.space)) }
+            let!(:binding2) { ServiceBinding.make(app: app, service_instance: ManagedServiceInstance.make(space: app.space)) }
+
             before do
-              allow_any_instance_of(ServiceBindingDelete).to receive(:delete).and_return([[StandardError.new('first'), StandardError.new('second')], []])
+              call_number = 0
+              allow_any_instance_of(V3::ServiceBindingDelete).to receive(:delete) do
```

Comment:
This `allow_any_instance` mock is a bit far away from the class under test. `AppDelete` includes `BindingDeleteMixin` which calls `DeleteServiceBindingFactory` which returns `V3::ServiceCredentialBindingDelete`/` V3::ServiceRouteBindingDelete` which then delegates to `V3::ServiceBindingDelete`, where the `delete` method is actually implemented. Whereas in the previous implementation we were calling the mocked class (`ServiceBindingDelete`) directly.

Is there something closer that we can mock? Maybe mock `delete_bindings` directly? That said, it might be a bit awkward to mock a private method included in the class under test. We don't have a great solution here, but this mock is kind of hard to understand without digging through multiple layers of dependencies.

Reply:
We agree it is not ideal. However if we mock `delete_bindings` in the `BindingDeleteMixin` we miss the fact that the underlying action throws but the `delete_bindings` method rescues the exceptions and returns a list of errors.
I tried to make it clearer just renaming things a bit. Hopefully that will at least remove the need to dig through multiple layers to understand it.